### PR TITLE
FIX: refactor and redesign connectivity

### PR DESCRIPTION
**Pull request by c9s**

### CODE REVIEW

File: pkg/types/connectivitygroup.go [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+
+const (
+	ConnectivityStateDisconnected ConnectivityState = -1
+	ConnectivityStateUnknown      ConnectivityState = 0
+	ConnectivityStateConnected    ConnectivityState = 1
+	ConnectivityStateAuthed       ConnectivityState = 2
+)
+
+func getConnState(con *Connectivity) ConnectivityState {
+	state := ConnectivityStateDisconnected
```
Comment (ycdesu) [inline pkg/types/connectivitygroup.go:34]: Should the default value be Unknown and add a disconnected else case?

Comment (c9s) [reply pkg/types/connectivitygroup.go:34]: i later figured out that we can start with disconnected state

File: pkg/types/connectivitygroup.go [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+		select {
+		case <-ctx.Done():
+			return false
+
+		case <-conn.connectedC:
+			continue
+
+		case <-conn.disconnectedC:
+			return true
+		}
```
Comment (gx578007) [inline pkg/types/connectivitygroup.go:192]: Is it possible to be blocked here forever? Any timeout required?

File: pkg/types/connectivitygroup.go [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+
+func (g *ConnectivityGroup) Add(con *Connectivity) {
+	g.mu.Lock()
+	g.connections = append(g.connections, con)
+	g.states[con] = getConnState(con)
+	g.sumState = sumStates(g.states)
+	g.setFlags(g.sumState)
+	g.mu.Unlock()
+
+	_con := con
```
[the remark this answers is not in the document]
Comment (c9s) [reply pkg/types/connectivitygroup.go:163]: copy 是故意留給 closure 用，我怕直接用 func argument 會有 side effect

File: pkg/types/connectivitygroup.go (hunk 2 above)
Comment (c9s) [reply pkg/types/connectivitygroup.go:192]: i need to remove the AnyDisconnected(), channel 會有 triggered once 的問題

**Review by gx578007 — APPROVED** [review]

**Review by narumiruna — APPROVED** [review]

**Pull request merged**